### PR TITLE
[2342] Add subjects filter page

### PR DESCRIPTION
**Pull request by faissalMT**
### Context
In order to progressively begin moving Find to Rails we will start by moving individual filter pages.

### Changes proposed in this pull request
Implement the subjects filter page with the same subject format as the C# implementation. This is currently using a hardcoded translation layer between C# and Rails subject ids.  
This does not include the filter for SEND courses, as this has been broken out into another card.

### CODE REVIEW

**Comment by faissalMT** [comment]
[[M] subjects filter: Add filter page](https://trello.com/c/F2j80t8R/2342-m-subjects-filter-add-filter-page)

File: app/helpers/csharp_rails_subject_conversion_helper.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      "36" => { csharp_id: "16", name: "German" },
+      "37" => { csharp_id: "20", name: "Italian" },
+      "38" => { csharp_id: "21", name: "Japanese" },
+      "39" => { csharp_id: "23", name: "Mandarin" },
+      "40" => { csharp_id: "41", name: "Russian" },
+      "41" => { csharp_id: "44", name: "Spanish" },
+      "43" => { csharp_id: "14", name: "Further education" },
+      "44" => { csharp_id: nil, name: "Humanities" },
+      "45" => { csharp_id: nil, name: "Balanced Science" },
+      "33" => { csharp_id: nil, name: "Modern Languages" },
```
Comment (defong) [inline app/helpers/csharp_rails_subject_conversion_helper.rb]: there is no reason for this

File: app/helpers/csharp_rails_subject_conversion_helper.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      "18" => { csharp_id: "8", name: "Design and technology" },
+      "19" => { csharp_id: "9", name: "Drama" },
+      "20" => { csharp_id: "11", name: "Economics" },
+      "21" => { csharp_id: "12", name: "English" },
+      "22" => { csharp_id: "15", name: "Geography" },
+      "23" => { csharp_id: "17", name: "Health and social care" },
+      "24" => { csharp_id: "18", name: "History" },
+      "25" => { csharp_id: "24", name: "Mathematics" },
+      "26" => { csharp_id: "27", name: "Music" },
+      "27" => { csharp_id: nil, name: "Philosophy" },
```
Comment (defong) [inline app/helpers/csharp_rails_subject_conversion_helper.rb]: there is no reason for this

**Comment by faissalMT** [comment]
@defong The fixes for hidden fields have already been merged upstream https://github.com/DFE-Digital/find-teacher-training/pull/52

**Comment by defong** [comment]
> @defong The fixes for hidden fields have already been merged upstream #52

@faissalMT rebase

File: app/helpers/csharp_rails_subject_conversion_helper.rb [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      "34" => { csharp_id: "13", name: "French" },
+      "35" => { csharp_id: "53", name: "English as a second or other language" },
+      "36" => { csharp_id: "16", name: "German" },
+      "37" => { csharp_id: "20", name: "Italian" },
+      "38" => { csharp_id: "21", name: "Japanese" },
+      "39" => { csharp_id: "23", name: "Mandarin" },
+      "40" => { csharp_id: "41", name: "Russian" },
+      "41" => { csharp_id: "44", name: "Spanish" },
+      "43" => { csharp_id: "14", name: "Further education" },
+      "44" => { csharp_id: nil, name: "Humanities" },
```
[the remark this answers is not in the document]
Comment (faissalMT) [reply app/helpers/csharp_rails_subject_conversion_helper.rb]: The list of C# subjects is smaller than the Rails one

File: app/views/result_filters/subjects/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<%= form_with(url: "/results/filter/subjects", method: :post) do |f| %>
```
Comment (defong) [inline app/views/result_filters/subjects/new.html.erb]:
```suggestion
<%= form_with(url: "/results/filter/subject", method: :post) do |f| %>
```

Comment (defong) [reply app/views/result_filters/subjects/new.html.erb]:
```suggestion
<%= form_with(url: subject_path, method: :post) do |f| %>
```

File: config/routes.rb [hunk 5]
```diff
@@ -16,6 +16,8 @@
   get "/results", to: "results#index", as: "results"
 
   scope module: "result_filters", path: "/results/filter" do
+    get "subjects", to: "subjects#new"
```
Comment (defong) [inline config/routes.rb]:
```suggestion
    get "subjects", to: "subject#new"
```

File: config/routes.rb [hunk 6]
```diff
@@ -16,6 +16,8 @@
...
+    post "subjects", to: "subjects#create"
```
Comment (defong) [inline config/routes.rb]:
```suggestion
    post "subject", to: "subjects#create"
```

File: app/controllers/result_filters/study_type_controller.rb [hunk 7]
```diff
@@ -5,11 +5,11 @@ class StudyTypeController < ApplicationController
     def new; end
 
     def create
-      if !(filter_params[:fulltime] == "True" || filter_params[:parttime] == "True")
+      if filter_params[:fulltime] == "True" || filter_params[:parttime] == "True"
```
Comment (defong) [inline app/controllers/result_filters/study_type_controller.rb]: not case sensitive

File: app/helpers/csharp_rails_subject_conversion_helper.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    end
+
+    raise "An unregistered subject id was specified: #{id}" if csharp_data.nil?
+
+    csharp_data[:csharp_id]
+  end
+
+  def csharp_rails_conversion_table
+    [{ csharp_id: "49", rails_id: "42", name: "Modern languages (other)" },
+     { csharp_id: nil, rails_id: "33", name: "Modern Languages" },
```
Comment (defong) [inline app/helpers/csharp_rails_subject_conversion_helper.rb]: do we need that?

File: spec/rails_helper.rb [hunk 9]
```diff
@@ -50,4 +50,17 @@
   config.filter_rails_from_backtrace!
   # arbitrary gems may also be filtered via:
   # config.filter_gems_from_backtrace("gem name")
+=begin
```
Comment (defong) [inline spec/rails_helper.rb]: whats this?

File: app/webpacker/styles/application.scss [hunk 10]
```diff
@@ -10,6 +9,7 @@
 @import "patterns/definition-list";
 @import "patterns/filters";
 @import "patterns/pagination";
+@import "accordion";
```
Comment (dankmitchell) [inline app/webpacker/styles/application.scss]: These styles  are already included on line  5 above

File: app/helpers/view_helper.rb [hunk 11]
```diff
@@ -14,4 +16,18 @@ def bat_contact_email_address
   def bat_contact_mail_to(name = nil, subject: nil, link_class: "govuk-link")
     mail_to bat_contact_email_address, name || bat_contact_email_address, subject: subject, class: link_class
   end
+
+  def subject_is_selected?(id:)
```
Comment (gpeng) [inline app/helpers/view_helper.rb]: These are specific to one page so I don't think they should go in the general `ViewHelper`. They should be in a specific `ResultFilters::SubjectHelper` or a view object used by that page.

File: app/views/result_filters/subject/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        <p class="govuk-body">Visit Get into Teaching to find out more about <%= govuk_link_to "postgraduate loans", "https://getintoteaching.education.gov.uk/funding-and-salary/overview/postgraduate-loans" %>, <%= govuk_link_to "bursaries and scholarships", "https://getintoteaching.education.gov.uk/funding-and-salary/overview" %>, and <%= govuk_link_to "funding by subject", "https://getintoteaching.education.gov.uk/funding-and-salary/overview/funding-by-subject" %>.</p>
+
+
+        <div class="accordion govuk-form-group" data-module="govuk-accordion">
+          <% @subject_areas.each do |subject_area| %>
+            <div class="govuk-accordion__section" data-qa="subject_area">
+              <div class="govuk-accordion__section-header">
+                <h2 class="govuk-accordion__section-heading" data-qa="subject_area__name">
+                  <button type="button" data-qa="subject_area__accordion_button" id="accordion-heading-<%= subject_area.typename %>" aria-controls="null-content-1" class="govuk-accordion__section-button" aria-expanded="<%= if subject_area_is_selected?(subject_area: subject_area) then "true" else "false" end %>">
```
Comment (gpeng) [inline app/views/result_filters/subject/new.html.erb]: `subject_area_is_selected?(subject_area: subject_area).to_s`

File: spec/features/result_filters/subjects_spec.rb [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+  let(:subject_areas_request) do
+    stub_api_v3_resource(
+      type: SubjectArea,
+      resources: subject_areas,
+      include: [:subjects],
+    )
+  end
+
+  let(:current_recruitment_cycle) { build(:recruitment_cycle) }
```
Comment (gpeng) [inline spec/features/result_filters/subjects_spec.rb]: There is a helper -> `stub_results_page_request` that does this down to line 76. We can only use it for now as once the results page has proper functionality we should probably be testing that the correct results appear in these features, in which case, we'll have to do proper set up. There's no point doing that now really as we're not testing the results anyway, just stubbing out call on results to stop it making the network request.

File: spec/helpers/view_helper_spec.rb [hunk 14]
```diff
@@ -20,4 +20,16 @@
       end
     end
   end
+
+  describe "#subject_is_selected?" do
```
Comment (gpeng) [inline spec/helpers/view_helper_spec.rb]: If we move the helpers these should move too.

File: spec/support/expect_url_with_query.rb [hunk 15]
```diff
@@ -0,0 +1,6 @@
+def expect_url_with_query(expected_path:, expected_query_params:)
```
Comment (gpeng) [inline spec/support/expect_url_with_query.rb]: There is already a helper that does this `spec/support/helpers.rb#163` `expect_page_to_be_displayed_with_query`

File: spec/helpers/view_helper_spec.rb (hunk 14 above)
Comment (dankmitchell) [reply spec/helpers/view_helper_spec.rb]: Done

File: app/views/result_filters/subject/new.html.erb (hunk 12 above)
Comment (gpeng) [reply app/views/result_filters/subject/new.html.erb]: Discussed offline. This causes a weird failure 🤷‍♂

**Pull request merged**